infobind/cache.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infobind
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultBindCacheSize = 5
+
+// bindData store the basic bind info and bindSql astNode.
+type bindData struct {
+	bindMeta
+	ast ast.StmtNode
+}
+
+// bindCache holds a bindDataMap, key:origin sql hash value: bindData slice.
+type bindCache struct {
+	Cache map[string][]*bindData
+}
+
+// Handler hold a atomic bindCache.
+type Handler struct {
+	bind atomic.Value
+}
+
+// HandleUpdater use to update the bindCache.
+type HandleUpdater struct {
+	parser         *parser.Parser
+	lastUpdateTime types.Time
+	ctx            sessionctx.Context
+	globalHandler  *Handler
+}
+
+type bindMeta struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	Status      int64 // If the bindMeta has been deleted, the status will be 0 or will be 1.
+	CreateTime  types.Time
+	UpdateTime  types.Time
+	Charset     string
+	Collation   string
+}
+
+// NewHandleUpdater create a new HandleUpdater.
+func NewHandleUpdater(ctx sessionctx.Context, handler *Handler, parser *parser.Parser) *HandleUpdater {
+	return &HandleUpdater{
+		globalHandler: handler,
+		parser:        parser,
+		ctx:           ctx,
+	}
+}
+
+// NewHandler create a Handler with a bindCache.
+func NewHandler() *Handler {
+	handler := &Handler{}
+	return handler
+}
+
+// Get get bindCache from a Handler.
+func (h *Handler) Get() *bindCache {
+	bc := h.bind.Load()
+
+	if bc != nil {
+		return bc.(*bindCache)
+	}
+
+	return &bindCache{
+		Cache: make(map[string][]*bindData, defaultBindCacheSize),
+	}
+}
+
+// LoadDiff use to load new bind info to bindCache bc.
+func (h *HandleUpdater) loadDiff(sql string, bc *bindCache) error {
+	recordSets, err := h.ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := recordSets[0]
+	defer terror.Call(rs.Close)
+
+	chkBatch := rs.NewRecordBatch()
+	for {
+		err = rs.Next(context.TODO(), chkBatch)
```

Comment:
Can we use context.Backgroud() here?

Reply:
yes,but the result is the same between context.TODO with context.Background.